middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,302 @@
+from abc import ABC
+from abc import abstractclassmethod
+from abc import abstractmethod
+
+from mypy.types import Any
+
+from construct import Byte
+from construct import Bytes
+from construct import BytesInteger
+from construct import Container
+from construct import Enum
+from construct import GreedyBytes
+from construct import Int8ub
+from construct import PascalString
+from construct import Struct
+from construct import VarInt
+
+from golem_messages.cryptography import ecdsa_sign
+from golem_messages.cryptography import ecdsa_verify
+from golem_messages.exceptions import InvalidSignature
+from golem_messages.message.base import Message as BaseGolemMessage
+
+from .constants import FRAME_SEPARATOR
+from .constants import FRAME_SEPARATOR_LENGTH
+from .constants import PayloadType
+from .registry import PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS
+from .registry import register
+from .exceptions import MiddlemanProtocolError
+
+
+class AbstractMiddlemanMessage(ABC):
+    """
+    Abstract class for inheriting by Middleman protocol messages.
+
+    Contains base functionality like:
+    - serialization,
+    - deserialization,
+    - frame validation.
+
+    Subclasses must implement:
+    - parsing payload to bytes on serialization,
+    - parsing serialized bytes payload back to original format on deserialization,
+    - validation of original payload before serialization.
+
+    """
+
+    __slots__ = [
+        'payload',
+        'payload_type',
+        'request_id',
+    ]
+
+    def __init__(self, payload: Any, request_id: int) -> None:
+        self._validate_request_id(request_id)
+        self.request_id = request_id
+
+        self._validate_payload(payload)
+        self.payload = self._serialize_payload(payload)
+
+    @classmethod
+    @abstractclassmethod
+    def _deserialize_payload(cls, payload: bytes):
+        pass
+
+    @abstractmethod
+    def _serialize_payload(self, payload: Any):
+        pass
+
+    @abstractmethod
+    def _validate_payload(self, payload: Any):
+        pass
+
+    @classmethod
+    def get_frame_format(cls) -> Struct:
+        """ Returns Struct object containing frame structure. """
+        return Struct(
+            frame_signature=Bytes(65),
+            signed_part_of_the_frame=Struct(
+                request_id=BytesInteger(4),
+                payload_type=Enum(Byte, PayloadType),
+                payload_length=BytesInteger(4),
+                payload=GreedyBytes,
+            ),
+        )
+
+    @classmethod
+    def add_frame_separator(cls, raw_frame: bytes) -> bytes:
+        """ Adds frame separator to raw frame. """
+        assert isinstance(raw_frame, bytes)
+
+        return raw_frame + FRAME_SEPARATOR
```

Comment:
I would make it optional, as a parameter to `.serialize`, default can be `False`. How about that ?

Reply:
I don't think that's a good solution because it just violates separation of concerns, even if it's optional. The frame is something that's between two separators. The function responsible for parsing the stream detects separators, cuts out the serialized frame from between them, unescapes it and returns as a string of bytes. At this point this string of bytes has nothing to do with the separator.

It's just an implementation detail from the lower level of abstraction. It's as if you were submitting multiple Golem messages to Concent in a multipart HTTP request and then included multipart separators in those messages. This seems wrong because there would be no separators it they were not sent over HTTP.

Reply:
That is why it is `False` by default in both `serialize` and `deserialize`. Please consider that is is library and it might be used in a different way than just planned. This function along with parameter which allows calling it during `serialize/deserialize` allows that.